src/spec/response.ts

```diff
@@ -1,50 +1,80 @@
-import { Message, MessageType  } from './message';
+import { Message, Attachment  } from './message';
 import { Translator } from '../translator';
 import * as program from 'commander';
 
 const wildcardRegex: RegExp = /<\*>/g;
 const wordRegex: RegExp = /<WORD>/g;
 const numberRegex: RegExp = /<NUMBER>/g;
-const imageRegex: RegExp = /<IMAGE>/g;
-const cardsRegex: RegExp = /<CARDS>/g;
 const regexRegex: RegExp = /<\((.+?)\)>/g;
 
+// Matches an attachment wildcard (e.g. <IMAGE>) at the _end_ of responseData.
+//
+// The matched part is just the word in the brackets ('IMAGE'), which can be
+// directly compared against the Attachment enum (from which it was generated).
+const attachmentAlternatives = Object.values(Attachment).join('|');
+const attachmentsRegex = new RegExp(`\\s*<(${attachmentAlternatives})>$`);
+
 export class Response {
-  responseType: MessageType;
+  attachments: Attachment[];
   private textMatchChecker: RegExp;
+  private bodyText: string;
   original: string;
 
   constructor(responseData: string) {
+    this.attachments = [];
     this.original = responseData.trim();
-    switch (this.original) {
-      case '<IMAGE>':
-        this.responseType = MessageType.Image;
-        break;
-      case '<CARDS>':
-        this.responseType = MessageType.Card;
-        break;
-      default:
-        this.responseType = MessageType.Text;
-        this.textMatchChecker = new RegExp(Response.transformTags(this.original));
+    this.bodyText = this.original;
+
+    // Construct a list of attachments from the attachment wildcards at the
```

Comment:
I like this solution, it's nice. Though I wonder why you constrain it to happening at the end of the line? Seems a bit artificial to me?

Reply:
In principle they could go anywhere, but I’m not sure what the use case would be. For example:

```YAML
Dialogue:
  - Human: How is my campaign doing?
  - Bot: According to <IMAGE> and <IMAGE>, it’s going great!
```

This doesn’t make any sense in bot framework _per se_, because it doesn’t define a mechanism for embedding attachments inline with the text. Presumably you would expect `attachments` to be `[Attachment.Image, Attachment.Image]`, but is `text` supposed to be `'According to'`, or `'According to and , it’s going great!`, or something else?

You can imagine a client which does support this (e.g. one that uses `$0`, `$1`, … to reference attachments in the `attachments` array), in which case you could check that `text === 'According to $0 and $1, it’s going great!'`, and that `attachments[0] === Attachment.Image` and `attachments[1] === Attachment.Image`; but in the absence of an established convention I’m not sure how useful this would be.

Reply:
Actually you are right that the botbuilder platform specifically limits this ordering. However, for other platforms like Slack, it is allowed to have intermingled text + attachments, so for that reason it might be good to allow the tags anywhere, and just concatenate the text.

It could lead to this edge case where someone sends

Hi <IMAGE> there <IMAGE>

and it validates against:

HI there <IMAGE> <IMAGE>

However at the same time, we could introduce this fix as a separate PR as you don't have any example connector right now, so it would be hard to test. For now we can concatenate and it will work for most practical use cases I think.

So to summarize I think for now a good compromise is to just drop the ordering condition (remove the `$` on the regex parser).